### PR TITLE
Document Reusable Fields

### LINKED ISSUE #152: abapLanguageVersion - document design decisions
see https://github.com/SAP/abap-file-formats/issues/44

### PR DESCRIPTION
**Pull request by schneidermic0**
Closes #152.

### CODE REVIEW

File: doc/specification.md [hunk 1]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+## Reusable Fields in JSON files
+
+In this section fields are described which can be reused by all file formats.
+
+The types for these reusable are defined in interface [`ZIF_AFF_TYPES_V1`](file-formats/typesUsedForAll/zif_aff_types_v1.intf.abap).
+
+### Schema
+
+The field `$schema` is a meta field and defines the format and the format version of the json content.
+Basically, it is  needed to allow the incompatible evolution of an content type.
```
Comment (wurzka) [inline doc/specification.md:350]:
```suggestion
Basically, it is needed to allow the incompatible evolution of a content type.
```

File: doc/specification.md [hunk 2]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+Translations of the texts shall be stored in separate files.
+
+The original language is not specified for sub objects. Sub objects have the same original version as their main objects.
+
+### ABAP Language Version
+
+The field `abapLanguageVersion` specifies the ABAP language version of an object.
+
+
+With the ABAP language version, it is possible to specify which ABAP language   elements and which other objects you can use in your object.
```
Comment (wurzka) [inline doc/specification.md:386]:
```suggestion
With the ABAP language version, it is possible to specify which ABAP language elements and which other objects you can use in your object.
```

**Review by wurzka — CHANGES_REQUESTED** [review]
Just some small changes

File: doc/specification.md [hunk 3]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
   ]
 }
 ```
+## Reusable Fields in JSON files
```
Comment (albertmink) [inline doc/specification.md:341]:
```suggestion
## Reusable Fields in JSON Files
```

File: doc/specification.md [hunk 4]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+
+In this section fields are described which can be reused by all file formats.
```
Comment (albertmink) [inline doc/specification.md:343]:
```suggestion
This section describes fields that can be reused by all file formats.
```

File: doc/specification.md [hunk 5]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+
+The types for these reusables are defined in the interface [`ZIF_AFF_TYPES_V1`](file-formats/typesUsedForAll/zif_aff_types_v1.intf.abap).
+
+### Schema
+
+The field `$schema` is a meta field and defines the format and the format version of the json content.
```
Comment (albertmink) [inline doc/specification.md:349]:
```suggestion
The field `$schema` is a meta field and defines the format and the format version of the JSON content.
```

File: doc/specification.md [hunk 6]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+Basically, it is needed to allow the incompatible evolution of a content type.
+
+`$schema` specifies the location of the json schema with an URI.
```
Comment (albertmink) [inline doc/specification.md:352]:
```suggestion
The field `$schema` specifies the location of the JSON schema with an URI.
```

File: doc/specification.md [hunk 7]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
 
 ---
+
+### Description
+
+The field `description` contains the description of the object.
+
+### Original Language
+
+The field `originalLanguage` stores the information about the original language of the the object.
```
Comment (albertmink) [inline doc/specification.md:366]:
```suggestion
The field `originalLanguage` stores the information about the original language of the object.
```

File: doc/specification.md [hunk 8]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+
+### ABAP Language Version
+
+The field `abapLanguageVersion` specifies the ABAP language version of an object.
+
+
+With the ABAP language version, it is possible to specify which ABAP language elements and which other objects you can use in your object.
+It does not specify the compatibility with a specific release (like SAP_BASIS 7.55).
+The ABAP language version `standard` specifies no limitation with regards to the usage of ABAP language elements or other objects.
+The ABAP language versions `cloudDevelopment` and `keyUser` are a subset of ABAP language version standard and you can use only a subset of ABAP language  elements and released objects.
```
Comment (albertmink) [inline doc/specification.md:389]:
```suggestion
The ABAP language versions `cloudDevelopment` and `keyUser` are a subset of ABAP language version standard and you can use only a subset of ABAP language elements and released objects.
```

File: doc/specification.md [hunk 9]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+For more details you can refer to the [ABAP keyword documentation](https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/index.htm).
+
+The field `abapLanguageVersion` is optional. Allowed values are `standard`, `cloudDevelopment` or `keyUser`. The default value is `standard`.
+
+Implementations of ABAP file format can decide whether they serialize the ABAP language version.
```
Comment (albertmink) [inline doc/specification.md:394]:
```suggestion
Implementations of ABAP file formats can decide whether they serialize the ABAP language version.
```

File: doc/specification.md (hunk 4 above)
Comment (jhunni) [inline doc/specification.md]:
```suggestion
In this section, fields are described which can be reused by all file formats.
```

**Review by albertmink — COMMENTED** [review]
Some polishing

File: doc/specification.md (hunk 7 above)
Comment (jhunni) [inline doc/specification.md]:
```suggestion
The field `originalLanguage` stores the information about the original language of the object.
```

File: doc/specification.md [hunk 10]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+The field `description` contains the description of the object.
+
+### Original Language
+
+The field `originalLanguage` stores the information about the original language of the the object.
+
+The original language is specified with [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) language codes.
+
+---
+**Note**: It has to be decided whether the language code is specified upper case (like in ABAP) or lower case as in the ISO 639-1 specification.
```
Comment (jhunni) [inline doc/specification.md:371]:
```suggestion
**Note**: It has to be decided whether the language code is specified in upper case (like in ABAP) or lower case as in the ISO 639-1 specification.
```

File: doc/specification.md [hunk 11]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+Refer to issue [#34](https://github.com/SAP/abap-file-formats/issues/53).
+
+---
+
+All translatable texts in the object shall be maintained in its original language.
```
Comment (jhunni) [inline doc/specification.md:376]:
```suggestion
All translatable texts in the object shall be maintained in their original language.
```

File: doc/specification.md (hunk 8 above)
Comment (jhunni) [inline doc/specification.md]:
```suggestion
The ABAP language versions `cloudDevelopment` and `keyUser` are a subset of the ABAP language version standard and you can use only a subset of ABAP language  elements and released objects.
```

File: doc/specification.md [hunk 12]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+### ABAP Language Version
+
+The field `abapLanguageVersion` specifies the ABAP language version of an object.
+
+
+With the ABAP language version, it is possible to specify which ABAP language elements and which other objects you can use in your object.
+It does not specify the compatibility with a specific release (like SAP_BASIS 7.55).
+The ABAP language version `standard` specifies no limitation with regards to the usage of ABAP language elements or other objects.
+The ABAP language versions `cloudDevelopment` and `keyUser` are a subset of ABAP language version standard and you can use only a subset of ABAP language  elements and released objects.
+For more details you can refer to the [ABAP keyword documentation](https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/index.htm).
```
Comment (jhunni) [inline doc/specification.md:390]:
```suggestion
For more details, you can refer to the [ABAP keyword documentation](https://help.sap.com/doc/abapdocu_latest_index_htm/latest/en-US/index.htm).
```

File: doc/specification.md [hunk 13]
```diff
@@ -332,7 +338,60 @@ This leads to the following generated JSON schema:
...
+
+The field `abapLanguageVersion` is optional. Allowed values are `standard`, `cloudDevelopment` or `keyUser`. The default value is `standard`.
+
+Implementations of ABAP file format can decide whether they serialize the ABAP language version.
+
+During deserialization, ABAP systems might change the ABAP language version needed in the specific context.
+E.g., SAP BTP, ABAP environment systems set the ABAP language version to `cloudDevelopment` or systems which do not support the ABAP language version might set it to `standard`.
```
Comment (jhunni) [inline doc/specification.md:397]:
```suggestion
E.g., SAP BTP, ABAP environment systems set the ABAP language version to `cloudDevelopment`, while systems which do not support the ABAP language version might set it to `standard`.
```

**Review by jhunni — APPROVED** [review]

**Review by wurzka — APPROVED** [review]
Looks good to me :)

**Review by albertmink — APPROVED** [review]
Look's good to me.

**Pull request merged**